buck2: update benchmark (#4175)
buildbuddy-io/buildbuddy: website/blog/buck2.md

     name = "xyz",
+    srcs = [
+        ":z_{}".format(n)
+        for n in SIZE_RANGE
+    ],
     outs = ["xyz.txt"],
     # cat $(location :z_0) $(location :z_1) ... $(location :z_999) > $@,
     cmd = SLEEP_CMD + "cat " + " ".join([
         "$(location :z_{})".format(n)
         for n in SIZE_RANGE
     ]) + " > $@",
-    tools = [
-        ":z_{}".format(n)
-        for n in SIZE_RANGE
-    ],
 )
```

Using `tools` would trigger Bazel to configure twice the number of actions underneath due to Bazel's special treament for tool dependencies.

And fewer actions means faster build.
For example, here is the new benchmark when build and test everything with Bazel.

```bash
> hyperfine --prepare 'bazel clean' --warmup 1 'bazel test --config=local //...'
Benchmark 1: bazel test --config=local //...
  Time (mean ± σ):     42.580 s ±  1.881 s    [User: 0.039 s, System: 0.048 s]
  Range (min … max):   38.127 s … 45.369 s    10 runs
```

So I reran the benchmark and here are the new timings.

| Tools       |          Local Build | RBE (with Remote Cache) | RBE (without Remote Cache) |
| :---------- | -------------------: | ----------------------: | -------------------------: |

> | buck2 (old) | ` 1.175 s ± 2.183 s` |    ` 7.937 s ± 0.487 s` |     `203.431 s ± 29.882 s` |

Reviewer: Is the local build number here correct? 1s seems like a big difference

Author: I copied and pasted with 1 missed character. It's `41`, not `1` 😵‍💫

| bazel (old) | `65.631 s ±  0.460 s` |     ` 9.798 s ±  0.515 s` |       `106.062 s ± ‎ 3.106 s` |

> | bazel (new) | `42.580 s ±  1.881 s` |     `10.047 s ±  0.517 s` |       ` 73.009 s ± ‎ 2.975 s` |

Author: Mildly interesting: there are 2 "empty" Unicode characters here, right in front of `3.106 s` and `2.975 s` to help with the indentation on the blog.

Github does not render them but they are there.

This means that if setup correctly, Bazel performance is quite competitive with Buck2.
Differences in a few seconds could be negligible for most use cases.

Tobias also noted that my current Bazel configuration is uploading Build Events, which may incur additional network latency.
This could be further be improved by moving the Build Event upload to a background task and defer build artifact downloads.

Here is a quick comparision with RBE and Remote Cache on different modes:

```bash
> hyperfine --prepare 'bazel clean' \
            --warmup 1 \
            'bazel test --config=remote //...' \
            'bazel test --config=remote //... --bes_upload_mode=fully_async' \
            'bazel test --config=remote //... --bes_upload_mode=fully_async --remote_download_toplevel' \
            'bazel test --config=remote //... --bes_upload_mode=fully_async --remote_download_minimal'

Benchmark 1: bazel test --config=remote //...
  Time (mean ± σ):      9.460 s ±  0.540 s    [User: 0.023 s, System: 0.031 s]
  Range (min … max):    8.473 s … 10.107 s    10 runs

> Benchmark 2: bazel test --config=remote //... --bes_upload_mode=fully_async

Reviewer 2: Is the buck2 build configured to use its BES equivalent? If not, I'm not sure it makes sense to compare against Bazel with BES enabled

Author: Updated. PTAL 🙏

And fewer actions means a faster build.

So I reran the benchmark and here are the new results:

| Tools            |          Local Build | RBE (with Remote Cache) | RBE (without Remote Cache) |
| :--------------- | -------------------: | ----------------------: | -------------------------: |
| buck2            | `41.175 s ± 2.183 s` |    ` 7.937 s ± 0.487 s` |     `203.431 s ± 29.882 s` |
| bazel w. `tools` | `65.631 s ± 0.460 s` |    ` 9.798 s ± 0.515 s` |    `106.062 s ± ‎ 3.106 s` |
| bazel w. `srcs`  | `42.580 s ± 1.881 s` |    `10.047 s ± 0.517 s` |    ` 73.009 s ± ‎ 2.975 s` |

This means that if setup correctly, Bazel performance is quite competitive with Buck2.
Differences of a few seconds could be negligible for most use cases.

Another point of improvement is using the default BuildBuddy setting, we are uploading Bazel's Build Events to BuildBuddy on each build.
With Buck2, there is an equivalent Buck2 Event system.
However, as noted in the blog post, Buck2 Event system is not yet compatible with systems outside of Meta so it's not possible to test with it just yet.

> So to achieve a more accurate test result, we could move Bazel's Build Event upload to a separate config as follow.

Reviewer 2: ```suggestion
So to achieve a more accurate test result, we could move Bazel's Build Event upload to a separate config as follows:
```

-build:remote --bes_backend=grpcs://remote.buildbuddy.io
 build:remote --remote_cache=grpcs://remote.buildbuddy.io
 build:remote --remote_timeout=3600
 build:remote --remote_executor=grpcs://remote.buildbuddy.io

+build:remote-w-bes --config=remote
+build:remote-w-bes --bes_results_url=https://app.buildbuddy.io/invocation/
+build:remote-w-bes --bes_backend=grpcs://remote.buildbuddy.io
```

How much of an improvement is it?

```bash
> hyperfine --prepare 'bazel clean' \
            --warmup 1 \
            'bazel test --config=remote //...' \
            'bazel test --config=remote-w-bes //...'

Benchmark 1: bazel test --config=remote //...
  Time (mean ± σ):      6.208 s ±  0.444 s    [User: 0.022 s, System: 0.031 s]
  Range (min … max):    5.392 s …  6.860 s    10 runs

Benchmark 2: bazel test --config=remote-w-bes //...
  Time (mean ± σ):     10.299 s ±  0.300 s    [User: 0.025 s, System: 0.036 s]
  Range (min … max):    9.811 s … 10.739 s    10 runs

Summary
  bazel test --config=remote //... ran
    1.66 ± 0.13 times faster than bazel test --config=remote-w-bes //...
```

#### Putting all new improvements together

> Here is the new benchmark results

Reviewer 2: ```suggestion
Here are the new benchmark results:
```

| Tools                         |          Local Build | RBE (with Remote Cache) | RBE (without Remote Cache) |
| :---------------------------- | -------------------: | ----------------------: | -------------------------: |
| buck2                         | `41.175 s ± 2.183 s` |    ` 7.937 s ± 0.487 s` |     `203.431 s ± 29.882 s` |
| bazel w. `tools` (old)        | `65.631 s ± 0.460 s` |    ` 9.798 s ± 0.515 s` |    `106.062 s ± ‎ 3.106 s` |
| bazel w. `srcs`               | `42.580 s ± 1.881 s` |    `10.047 s ± 0.517 s` |    ` 73.009 s ± ‎ 2.975 s` |
| bazel<br/>w. `srcs`, no `BES` | `40.455 s ± 2.275 s` |    ` 6.208 s ± 0.444 s` |    ` 70.788 s ± ‎ 4.408 s` |

> This shows that when configured correctly, Bazel performance is very competitive comparing to Buck2.
> And that Buck2 still has some rooms for growth in areas such as Remote Build Execution.

Reviewer 2: ```suggestion
This shows that when configured correctly, Bazel's performance is very competitive with Buck2's,
and that Buck2 still has some room for growth in areas such as Remote Build Execution.
```

             ":x_{}".format(n),
             ":y_{}".format(n),
         ],
+        outs = ["z_{}.txt".format(n)],
+        cmd = SLEEP_CMD + "cat $(location :x_{n}) $(location :y_{n}) > $@".format(n = n),
     )
     for n in SIZE_RANGE
 ]

 genrule(
     name = "xyz",
+    srcs = [
+        ":z_{}".format(n)
+        for n in SIZE_RANGE
+    ],
     outs = ["xyz.txt"],
     # cat $(location :z_0) $(location :z_1) ... $(location :z_999) > $@,
     cmd = SLEEP_CMD + "cat " + " ".join([
         "$(location :z_{})".format(n)
         for n in SIZE_RANGE
     ]) + " > $@",
-    tools = [
-        ":z_{}".format(n)
-        for n in SIZE_RANGE
-    ],
 )
```

Using `tools` would trigger Bazel to configure twice the number of actions underneath due to Bazel's special treament for tool dependencies.

And fewer actions means faster build.
For example, here is the new benchmark when build and test everything with Bazel.

```bash
> hyperfine --prepare 'bazel clean' --warmup 1 'bazel test --config=local //...'
Benchmark 1: bazel test --config=local //...
  Time (mean ± σ):     42.580 s ±  1.881 s    [User: 0.039 s, System: 0.048 s]
  Range (min … max):   38.127 s … 45.369 s    10 runs

> ```

Author: ack. Will remove!

| :--------------- | -------------------: | ----------------------: | -------------------------: |
| buck2            | `41.175 s ± 2.183 s` |    ` 7.937 s ± 0.487 s` |     `203.431 s ± 29.882 s` |
| bazel w. `tools` | `65.631 s ± 0.460 s` |    ` 9.798 s ± 0.515 s` |    `106.062 s ± ‎ 3.106 s` |
| bazel w. `srcs`  | `42.580 s ± 1.881 s` |    `10.047 s ± 0.517 s` |    ` 73.009 s ± ‎ 2.975 s` |

This means that if setup correctly, Bazel performance is quite competitive with Buck2.
Differences in a few seconds could be negligible for most use cases.

Tobias also noted that my current Bazel configuration is uploading Build Events, which may incur additional network latency.
This could be further be improved by moving the Build Event upload to a background task and defer build artifact downloads.

Here is a quick comparision with RBE and Remote Cache on different modes:

```bash
> hyperfine --prepare 'bazel clean' \
            --warmup 1 \
            'bazel test --config=remote //...' \
            'bazel test --config=remote //... --bes_upload_mode=fully_async' \
            'bazel test --config=remote //... --bes_upload_mode=fully_async --remote_download_toplevel' \

>             'bazel test --config=remote //... --bes_upload_mode=fully_async --remote_download_minimal'

Author: agree

  Time (mean ± σ):      7.091 s ±  0.137 s    [User: 0.017 s, System: 0.035 s]
  Range (min … max):    6.893 s …  7.314 s    10 runs

Benchmark 3: bazel test --config=remote //... --bes_upload_mode=fully_async --remote_download_toplevel
  Time (mean ± σ):      5.733 s ±  0.294 s    [User: 0.016 s, System: 0.038 s]
  Range (min … max):    5.434 s …  6.261 s    10 runs

Benchmark 4: bazel test --config=remote //... --bes_upload_mode=fully_async --remote_download_minimal
  Time (mean ± σ):      5.659 s ±  0.189 s    [User: 0.016 s, System: 0.040 s]
  Range (min … max):    5.477 s …  6.140 s    10 runs

Summary
  bazel test --config=remote //... --bes_upload_mode=fully_async --remote_download_minimal ran
    1.01 ± 0.06 times faster than bazel test --config=remote //... --bes_upload_mode=fully_async --remote_download_toplevel
    1.25 ± 0.05 times faster than bazel test --config=remote //... --bes_upload_mode=fully_async
    1.67 ± 0.11 times faster than bazel test --config=remote //...
```

Buck2 does also come with "Deferred Materialization", the equivalent of Bazel's build-without-the-bytes.
However because the files we are generating are relatively lightweighted,

> I decided to not use these levers as part of the Buck2 benchmark.

Author: Agree. I will remove mention of both `bes_upload_mode` and `remote_download_{minimum,toplevel}`.

The following diff change to move `bes_backend` to a different Bazel config should be sufficient to compare against Buck2. So this section is a bit redundant overall.

  Time (mean ± σ):      6.208 s ±  0.444 s    [User: 0.022 s, System: 0.031 s]
  Range (min … max):    5.392 s …  6.860 s    10 runs

Benchmark 2: bazel test --config=remote-w-bes //...
  Time (mean ± σ):     10.299 s ±  0.300 s    [User: 0.025 s, System: 0.036 s]
  Range (min … max):    9.811 s … 10.739 s    10 runs

Summary
  bazel test --config=remote //... ran
    1.66 ± 0.13 times faster than bazel test --config=remote-w-bes //...
```

#### Putting all new improvements together

Here is the new benchmark results

| Tools                         |          Local Build | RBE (with Remote Cache) | RBE (without Remote Cache) |
| :---------------------------- | -------------------: | ----------------------: | -------------------------: |
| buck2                         | `41.175 s ± 2.183 s` |    ` 7.937 s ± 0.487 s` |     `203.431 s ± 29.882 s` |
| bazel w. `tools`              | `65.631 s ± 0.460 s` |    ` 9.798 s ± 0.515 s` |    `106.062 s ± ‎ 3.106 s` |
| bazel w. `srcs`               | `42.580 s ± 1.881 s` |    `10.047 s ± 0.517 s` |    ` 73.009 s ± ‎ 2.975 s` |
| bazel<br/>w. `srcs`, no `BES` | `40.455 s ± 2.275 s` |    ` 6.208 s ± 0.444 s` |    ` 70.788 s ± ‎ 4.408 s` |

> This shows that when configured correctly, Bazel performance could be very competitive comparing to Buck2.

Reviewer 3: nit: you might consider changing "could be" to "is" or turn the sentence around?

Author: Good point.

I think this is something definitively true, at least for now, so I don't mind changing.
